src/Duration.php

```diff
@@ -0,0 +1,187 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Spatie\Period;
+
+use DateInterval;
+use DateTimeImmutable;
+
+final class Duration
+{
+    /** @var int */
+    private $seconds;
+
+    /** @var int */
+    private $precision;
+
+    private function __construct()
+    {
+    }
+
+    public static function make($value): self
+    {
+        if ($value instanceof Period) {
+            return self::fromPeriod($value);
+        }
+
+        if ($value instanceof PeriodCollection) {
+            return self::fromPeriodCollection($value);
+        }
+
+        if ($value instanceof DateInterval) {
+            return self::fromDateInterval($value);
+        }
+
+        $stringValue = (string) $value;
+
+        if (0 === strpos($stringValue, 'P')) {
+            return self::fromDateInterval(new DateInterval($stringValue));
+        }
+
+        return self::fromDateInterval(DateInterval::createFromDateString($stringValue));
+    }
+
+    public static function fromPeriod(Period $period): self
+    {
+        $start = $period->getIncludedStart();
+        $end = $period->getIncludedEnd();
+        $precision = $period->getPrecisionMask();
+
+        if ($precision === Precision::SECOND) {
+            $end = $end->add(new DateInterval('PT1S'));
+        } elseif ($precision === Precision::MINUTE) {
+            $end = $end->add(new DateInterval('PT1M'));
+        } elseif ($precision === Precision::HOUR) {
+            $end = $end->add(new DateInterval('PT1H'));
+        } elseif ($precision === Precision::DAY) {
+            $end = $end->add(new DateInterval('PT24H'));
+        } else {
+            throw new \InvalidArgumentException('A duration can only be determined up to a precision of Precision::DAY');
```

Comment:
Could we — this is just an idea — make this class even smarter, and allow for P1M, but _only_ if the start date is the first day of its month, and the end date is the last day of its month?

My reasoning is that if you're working with months, you probably don't care about smaller precision, so I feel that we need some way of handling this.

I'd like to hear your thoughts.

Reply:
That's the conundrum I can't quite resolve for myself at the moment - if a collection only consists of `($firstOfMonth, $lastOfMonth, Boundaries::EXCLUDE_NONE)`, this would probably work, but if I create such a collection, I already know that they all are full months and I probably don't have a need to compare their durations.

On the other hand, if we have `($firstOfMonth, $lastOfMonth)` with different boundaries, we would have `P1M` for each of them, but the actual durations would not be equal.

These are also only my thoughts, on the premise to avoid confusion for devs implementing the library asking why they get different durations. That being said, it's easy to just try it out by adding the behavior and then adding test cases to see how things work out ^^

Reply:
So the biggest part of the problem exists because of duration support on collections, correct?

Reply:
No :/ - if we would apply the `P1M` behavior, we would have the following:

```php
$a = Period::make('2018-01-01', '2018-01-31', Precision::MONTH); // P1M
$b = Period::make('2018-01-01', '2018-01-31', Precision::DAY); // P31D
$a->duration()->equals($b->duration()); // false
```

My expectation might differ from the general expectation here, but I would assume that `$a` and `$b` would be considered equal.

Reply:
Got it.. Ok let's, for a moment, go back to what we originally wanted to achieve: why the need for this duration class?

Reply:
😅 I’m not sure anymore 😅 I think it was because Period::length() returned ambiguous results depending on the given precision, so it was not possible to compare Periods with different precisions

Reply:
Can we, since durations are always saved as seconds, only compare these amounts of seconds to each other?

If we want to use durations to be able to compare two period lengths with different precisions, it makes sense to convert everything to the smallest precision unit, and make comparisons with those values.

Let's take your example:

```php
$a = Period::make('2018-01-01', '2018-01-31', Precision::MONTH); // P1M
$b = Period::make('2018-01-01', '2018-01-31', Precision::DAY); // P31D

$a->duration()->equals($b->duration());
```

The duration of `$a` is `2678400` seconds, if I'm not mistaken. `$b` has the same amount of seconds, regardless of its precision, hence their durations would be equal. We _should_ note in the readme that this doesn't necessarily mean that the two periods are equal, but that's the whole point of adding this functionality: being able to more loosely compare two periods.

Doesn't this eliminate the whole month issue?

Reply:
While trying to implement this (checking by the second), I figured out an example which I didn't get to work:

```php
$a = Period::make('2018-01-01', '2018-01-31', Precision::MONTH); // P1M
$b = Period::make('2018-02-01', '2018-02-28', Precision::MONTH); // P1M

$a->duration()->equals($b->duration()); // false
```

But I just had an idea on how to fix this, if it works out, the solution would be even nicer that the workarounds, I just need a little more time.

Reply:
Ok, looking forward to your feedback